recipes/mpfr/all/conanfile.py

```diff
@@ -29,148 +34,151 @@ class MpfrConan(ConanFile):
+            tc.extra_cflags.append(f"-I{self.build_folder}")
+        if is_msvc(self):
+            tc.extra_cflags.append("-FS")
+
+        env = tc.environment()
+        if is_msvc(self):
+            env.define("AR", "lib")
+            env.define("CC", "cl -nologo")
+            env.define("CXX", "cl -nologo")
+            env.define("LD", "link")
+            env.define("NM", "dumpbin -symbols")
+            env.define("OBJDUMP", ":")
+            env.define("RANLIB", ":")
+            env.define("STRIP", ":")
+        tc.generate(env) # Create conanbuild.conf
 
     def _extract_makefile_variable(self, makefile, variable):
-        makefile_contents = tools.load(makefile)
-        match = re.search("{}[ \t]*=[ \t]*((?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\\\"))*(?:\\\\\n(?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\"))*)*)\n".format(variable), makefile_contents)
+        makefile_contents = load(self, makefile)
+        match = re.search(f'{variable}[ \t]*=[ \t]*((?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\\\"))*(?:\\\\\n(?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\"))*)*)\n', makefile_contents)
         if not match:
-            raise ConanException("Cannot extract variable {} from {}".format(variable, makefile_contents))
+            raise ConanException(f"Cannot extract variable {variable} from {makefile_contents}")
         lines = [line.strip(" \t\\") for line in match.group(1).split()]
         return [item for line in lines for item in shlex.split(line) if item]
 
     def _extract_mpfr_autotools_variables(self):
-        makefile_am = os.path.join(self._source_subfolder, "src", "Makefile.am")
+        makefile_am = os.path.join(self.source_folder, "src", "Makefile.am") # src/src/Makefile.am
         makefile = os.path.join("src", "Makefile")
         sources = self._extract_makefile_variable(makefile_am, "libmpfr_la_SOURCES")
         headers = self._extract_makefile_variable(makefile_am, "include_HEADERS")
         defs = self._extract_makefile_variable(makefile, "DEFS")
         return sources, headers, defs
 
-    @contextlib.contextmanager
-    def _build_context(self):
-        if self.settings.compiler == "Visual Studio":
-            with tools.vcvars(self):
-                env = {
-                    "AR": "lib",
-                    "CC": "cl -nologo",
-                    "CXX": "cl -nologo",
-                    "LD": "link",
-                    "NM": "dumpbin -symbols",
-                    "OBJDUMP": ":",
-                    "RANLIB": ":",
-                    "STRIP": ":",
-                }
-                with tools.environment_append(env):
-                    yield
-        else:
-            yield
-
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
+        apply_conandata_patches(self)
+        with chdir(self, self.source_folder):
+            command = "./autogen.sh"
+            if os.path.exists(command):
+                self.run(command)
```

Comment:
```suggestion
```

Please stop trying doing this in all recipes. conan-center-index recipes don't support other source code than the ones listed in yml file, it has already been said I think.

Reply:
Hi @SpaceIm,
One of the key benefits of Conan 2.0 is its integrated support for Git and building directly from Git distributions. This is a critical feature for those of us that work for organizations that have to fix bugs in core math libraries like MPFR. As the CI shows, this recipe still works fine building from a tarball. Additionally, with Conan 2.0, and finally having gotten all of the autotconf tools to work with Conan 2.0 (see pending PRs for autoconf, autoconf-archive, and libtool), it is actually pretty straightforward to leverage autoconf with them.

Reply:
It shouldn't be accepted and @prince-chrismc I don't understand why you accept this modification while you have already said that conan team has decided that such logic shouldn't be allowed in conan-center recipes.
This code can hide issues in future releases because it requires libtool/automake/autoconf/m4 which are not in tool_requires.